agent/rpc/middleware/interceptors_test.go

```diff
@@ -0,0 +1,126 @@
+package middleware
+
+import (
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"github.com/stretchr/testify/require"
+)
+
+// obs holds all the things we want to assert on that we recorded correctly in our tests.
+type obs struct {
+	key    []string
+	start  time.Time
+	labels []metrics.Label
+}
+
+// recorderStore acts as an in-mem mock storage for all the RequestRecorder.Record() recorderFunc calls.
+type recorderStore struct {
+	lock  sync.Mutex
+	store map[string]obs
+}
+
+func (rs *recorderStore) put(key []string, o obs) {
+	rs.lock.Lock()
+	defer rs.lock.Unlock()
+
+	actualKey := strings.Join(append(key, o.labels[0].Value), "")
+	rs.store[actualKey] = o
+}
+
+func (rs *recorderStore) get(key []string) obs {
+	rs.lock.Lock()
+	defer rs.lock.Unlock()
+
+	actualKey := strings.Join(key, "")
+	return rs.store[actualKey]
+}
+
+var store = recorderStore{store: make(map[string]obs)}
+var simpleRecorderFunc = func(key []string, start time.Time, labels []metrics.Label) {
+	o := obs{key: key, start: start, labels: labels}
+
+	store.put(key, o)
+}
+
+type readRequest struct{}
+type writeRequest struct{}
+
+func (rr readRequest) IsRead() bool {
+	return true
+}
+
+func (wr writeRequest) IsRead() bool {
+	return false
+}
+
+// TestRequestRecorder_SimpleOK tests that the RequestRecorder can record a simple request.
+func TestRequestRecorder_SimpleOK(t *testing.T) {
```

Comment:
i can be persuaded to add some of these tests in a for loop but for now i thought it was easier to read like so.

Reply:
I agree those could be transformed to a table test, but I have no preference.

Reply:
I'd be inclined to keep the tests like so for now. Thinking that if we add more logic in `RequestRecorder` in terms of parsing the request, then we can refactor and use table tests for brevity.

Reply:
> I'd be inclined to keep the tests like so for now. Thinking that if we add more logic in `RequestRecorder` in terms of parsing the request, then we can refactor and use table tests for brevity.

I like this approach! No need for table tests if there isn't a large combination of inputs